src/core/ddsi/src/q_entity.c

```diff
@@ -3894,6 +4209,132 @@ static void add_proxy_builtin_endpoints(
 #undef LTE
 }
 
+#ifdef DDSI_INCLUDE_SECURITY
+
+void handshake_end_cb
+(
+  struct ddsi_handshake *handshake,
+  const struct ddsi_guid *lpguid,
+  const struct ddsi_guid *ppguid,
+  nn_wctime_t timestamp,
+  enum ddsi_handshake_state result)
+{
+  struct proxy_participant *proxypp;
+  struct participant *pp;
+  int64_t shared_secret;
+  int64_t permissions_hdl;
+  struct q_globals *gv;
+
+  DDSRT_UNUSED_ARG(timestamp);
+
+  assert(handshake);
+  assert(lpguid);
+  assert(ppguid);
+
+  gv = ddsi_handshake_get_globals(handshake);
+  assert(gv);
+
+  proxypp = ephash_lookup_proxy_participant_guid (gv->guid_hash, ppguid);
+  if (proxypp) {
+    pp = ephash_lookup_participant_guid (gv->guid_hash, lpguid);
+    if (pp) {
+      switch(result)
+      {
+      case STATE_HANDSHAKE_PROCESSED:
+        shared_secret = ddsi_handshake_get_shared_secret(handshake);
+        DDS_CLOG (DDS_LC_DISCOVERY, &gv->logconfig, "handshake (lguid="PGUIDFMT" rguid="PGUIDFMT") processed\n", PGUID (*lpguid), PGUID (*ppguid));
+        permissions_hdl = q_omg_security_check_remote_participant_permissions(gv->config.domainId, pp, proxypp);
+        if (permissions_hdl != 0) {
+          q_omg_security_register_remote_participant(pp, proxypp, shared_secret, permissions_hdl);
+          match_volatile_secure_endpoints(pp, proxypp);
+        }
+        break;
+
+      case STATE_HANDSHAKE_SEND_TOKENS:
+        DDS_CLOG (DDS_LC_DISCOVERY, &gv->logconfig, "handshake (lguid="PGUIDFMT" rguid="PGUIDFMT") send tokens\n", PGUID (*lpguid), PGUID (*ppguid));
+        q_omg_security_participant_send_tokens(pp, proxypp);
+        break;
+
+      case STATE_HANDSHAKE_OK:
+        DDS_CLOG (DDS_LC_DISCOVERY, &gv->logconfig, "handshake (lguid="PGUIDFMT" rguid="PGUIDFMT") succeeded\n", PGUID (*lpguid), PGUID (*ppguid));
+        update_proxy_participant_endpoint_matching(proxypp, pp);
+        ddsi_hsadmin_remove_from_fsm(proxypp->handshake_admin, handshake);
+        break;
+
+      case STATE_HANDSHAKE_TIMED_OUT:
+        DDS_CERROR (&gv->logconfig, "handshake (lguid="PGUIDFMT" rguid="PGUIDFMT") failed: (%d) Timed out\n", PGUID (*lpguid), PGUID (*ppguid), (int)result);
+        if (q_omg_participant_allow_unauthenticated(pp)) {
```

Comment:
This is intruiging: why even perform a handshake when it communication is allowed anyway? (I find the ``FAILED`` case even more intriguing. I should look it up in the spec, I know.)

Reply:
The spec mentions the AuthenticationTimeoutWait time when this time expires the validation of the authentication will fail. In the current implementation when the handshake times out and the local participant allows unauthenticated communication then the matching continues as if the remote was unauthenticated. However it is probably better to retry the handshake at a later time. This will be taken into account with the implementation of the handshake.